cudax/include/cuda/experimental/detail/hierarchy_dimensions.cuh

```diff
@@ -0,0 +1,811 @@
+      : levels(::cuda::std::forward<Levels>(ls)...)
+  {}
+
+  constexpr _CCCL_HOST_DEVICE hierarchy_dimensions_fragment(const ::cuda::std::tuple<Levels...>& ls) noexcept
+      : levels(ls)
+  {}
+  constexpr _CCCL_HOST_DEVICE hierarchy_dimensions_fragment(::cuda::std::tuple<Levels...>&& ls) noexcept
+      : levels(::cuda::std::forward<::cuda::std::tuple<Levels...>>(ls))
+  {}
+
+  constexpr _CCCL_HOST_DEVICE
+  hierarchy_dimensions_fragment(const BottomUnit& unit, const ::cuda::std::tuple<Levels...>& ls) noexcept
+      : levels(ls)
+  {}
+  constexpr _CCCL_HOST_DEVICE
+  hierarchy_dimensions_fragment(const BottomUnit& unit, ::cuda::std::tuple<Levels...>&& ls) noexcept
+      : levels(::cuda::std::forward<::cuda::std::tuple<Levels...>>(ls))
+  {}
+
+private:
+  // TODO is this useful enough to expose?
+  template <typename Unit, typename Level>
+  _CCCL_HOST_DEVICE constexpr auto levels_range() const noexcept
+  {
+    static_assert(has_level<Level, hierarchy_dimensions_fragment<BottomUnit, Levels...>>);
+    static_assert(has_level_or_unit<Unit, hierarchy_dimensions_fragment<BottomUnit, Levels...>>);
+    static_assert(detail::legal_unit_for_level<Unit, Level>);
+    return ::cuda::std::apply(detail::get_levels_range<Level, Unit, Levels...>, levels);
+  }
+
+public:
+  /**
+   * @brief Get a fragment of this hierarchy
+   *
+   * This member function can be used to get a fragment of the hierarchy its called on.
+   * It returns a hierarchy_dimensions_fragment that includes levels starting with the
+   * level specified in Level and ending with a level before Unit. Toegether with
+   * hierarchy_add_level function it can be used to create a new hierarchy that is a modification
+   * of an exsiting hierarchy.
+   * @par Snippet
+   * @code
+   * #include <cudax/hierarchy_dimensions.cuh>
+   *
+   * auto hierarchy = make_hierarchy(grid_dims(256), cluster_dims<4>(), block_dims<8, 8, 8>());
+   * auto fragment = hierarchy.fragment(block, grid);
+   * auto new_hierarchy = hierarchy_add_level(fragment, block_dims<128>());
+   * static_assert(new_hierarchy.count(thread, block) == 128);
+   * @endcode
+   * @par
+   *
+   * @tparam Unit
+   *   Type indicating what should be the unit of the resulting fragment
+   *
+   * @tparam Level
+   *   Type indicating what should be the top most level of the resulting fragment
+   */
+  template <typename Unit, typename Level>
+  _CCCL_HOST_DEVICE constexpr auto fragment(const Unit& = Unit(), const Level& = Level()) const noexcept
+  {
+    auto selected = levels_range<Unit, Level>();
+    // TODO fragment can't do constexpr queries because we use references here, can we create copies of the levels in
+    // some cases and move to the constructor?
+    return ::cuda::std::apply(
+      [](const auto&... levels) {
+        return hierarchy_dimensions_fragment<Unit, ::cuda::std::remove_reference_t<decltype(levels)>...>(levels...);
+      },
```

Comment:
We should definitely turn this into a function object, otherwise this will instantiate a full new type every time we instantiate this function

Reply:
I believe we still need to template that object on Unit, since its not present in selected, but at least its no longer templated on Level (unless I misunderstood the ask).
Moved it to a helper function object.